### PR TITLE
Feature/extensibility enhancements

### PR DESCRIPTION
**Pull request by gmarz**
* `ElasticClient` now implements `IHighLevelToLowLevelDispatcher` explicitly so the internal-ish dispatch methods are less exposed.  This required casting to `IHighLevelToLowLevelDispatcher` in order to access the dispatch methods, so I added an internal `Dispatcher { get { return this; } }` convenience property (much like we do with `Self` in the descriptors).

* Renamed `IDomainObject` to the more descriptive`INestSerializable`.

* Changed `CloneFrom()` back to internal since it doesn't need to be exposed.

### CODE REVIEW

File: src/Nest/Domain/ICustomJson.cs [hunk 1]
```diff
@@ -2,14 +2,12 @@
...
 	{
 		object GetCustomJson();
 	}
 
-	public interface IDomainObject
-	{
-		
-	}
+	// Any object that implements this will automatically be added to the serialization contract
+	public interface INestSerializable { }
```
Comment (Mpdreamz) [inline src/Nest/Domain/ICustomJson.cs]: lets make this a proper summary: `Any object that implements this interface will automatically have all the JsonProperties of all its implementing interfaces discovered.`

File: src/Nest/ElasticClient.cs [hunk 2]
```diff
@@ -172,17 +157,5 @@ private TRequest ForceConfiguration<TRequest>(TRequest request, Action<IRequestC
 			request.RequestConfiguration = configuration;
 			return request;
 		}
-
-		public static void Warmup()
```
Comment (Mpdreamz) [inline src/Nest/ElasticClient.cs]: This has always been public or not? if not we need to add it back in (sadly)

File: src/Elasticsearch.Net/Domain/Response/ElasticsearchResponse.cs [hunk 3]
```diff
@@ -19,11 +19,9 @@
 
 namespace Elasticsearch.Net
 {
-	//TODO document and possibly rename some of the properties
-
 	public static class ElasticsearchResponse
```
Comment (Mpdreamz) [inline src/Elasticsearch.Net/Domain/Response/ElasticsearchResponse.cs]: Add a todo to make this whole class `internal` in 2.0

File: src/Nest/ElasticClient.cs (hunk 2 above)
Comment (gmarz) [inline src/Nest/ElasticClient.cs]: It has been public, but I didn't remove it.  Just moved it up.

**Pull request merged**